packages/tx-construction/src/tx-builder/TxBuilder.ts

```diff
@@ -396,30 +404,51 @@ export class GenericTxBuilder implements TxBuilder {
               }
             }
 
-            const { body, hash, inputSelection, redeemers } = await initializeTx(
-              {
-                auxiliaryData,
-                certificates: this.partialTxBody.certificates,
-                collateralReturn,
-                collaterals,
-                customizeCb: this.#customizeCb,
-                handleResolutions: this.#handleResolutions,
-                inputs: new Set(this.#preSelectedInputs.values()),
-                options: {
-                  validityInterval: this.partialTxBody.validityInterval
-                },
-                outputs: new Set(this.partialTxBody.outputs || []),
-                proposalProcedures: this.partialTxBody.proposalProcedures,
-                redeemersByType: this.#knownRedeemers,
-                referenceInputs: new Set([...this.#referenceInputs.values()].map((utxo) => utxo[0])),
-                scriptIntegrityHash: hasPlutusScripts ? DUMMY_SCRIPT_DATA_HASH : undefined,
-                scriptVersions,
-                signingOptions: partialSigningOptions,
-                txEvaluator: this.#txEvaluator,
-                witness
+            const initializeTxProps = {
+              auxiliaryData,
+              certificates: this.partialTxBody.certificates,
+              collateralReturn,
+              collaterals,
+              customizeCb: this.#customizeCb,
+              handleResolutions: this.#handleResolutions,
+              inputs: new Set(this.#preSelectedInputs.values()),
+              options: {
+                validityInterval: this.partialTxBody.validityInterval
               },
-              dependencies
-            );
+              outputs: new Set(this.partialTxBody.outputs || []),
+              proposalProcedures: this.partialTxBody.proposalProcedures,
+              redeemersByType: this.#knownRedeemers,
+              referenceInputs: new Set([...this.#referenceInputs.values()].map((utxo) => utxo[0])),
+              scriptIntegrityHash: hasPlutusScripts ? DUMMY_SCRIPT_DATA_HASH : undefined,
+              scriptVersions,
+              signingOptions: partialSigningOptions,
+              txEvaluator: this.#txEvaluator,
+              witness
+            };
+
+            let initialTxResult;
+
+            try {
+              initialTxResult = await initializeTx(initializeTxProps, dependencies);
+            } catch (error) {
+              // Fallback to large first if we are not using the greedy selector
+              if (
+                usingGreedySelector ||
+                !(error instanceof InputSelectionError) ||
+                error.failure === InputSelectionFailure.UtxoBalanceInsufficient
+              ) {
+                throw error;
+              }
+
+              dependencies.inputSelector = new LargeFirstSelector({
```

Comment:
Ideally fallback inputSelector would be a separate dependency, but this is just a minor nit